### ISSUE TITLE
[bug] Access Policy: Resources

### ISSUE THREAD

**Issue by Ruslan-Aleev** [modxcms/revolution#14479]
### Summary
Recently we tested all the settings in the **Access Policy** and grouped them into logical groups, but almost every setting has an inaccuracy or even a bug.

**Guys, we need help in testing these problems, both from the MODX-user and from the MODX-components.**

_"Resources" Group:_
- **new_document_in_root** - Prevent creating resource at the root. There are no comments.
- **new_document** - Prevent creating resource. At the same time, the resource icon disappears from the quick menu, but the remaining icons are a symlink, a weblink is available, which is strange.
- **new_symlink** - Prevent creating symlink. The quick creation icon in the tree disappears. At the same time, you can create symlink by creating a resource and choosing Resource Type in the resource settings.
- **new_weblink** - Prevent creating weblink. The quick creation icon in the tree disappears. At the same time, you can create weblink by creating a resource and choosing Resource Type in the resource settings.
- **new_static_resource** - Prevent creating static resource. The quick creation icon in the tree disappears. At the same time, you can create static resources by creating a resource and choosing Resource Type in the resource settings.
- **edit_document** - Prevent editing resource. There are no comments.
- **save_document** - Prevent saving resources. And the button in the form is also locked. There are no comments here.
- **delete_document** - To delete or remove resource. Description Question: Why delete and remove, is it not the same thing?
- **resource_duplicate** - Prevent creating copy of the resource. There are no comments here.
- **resource_quick_create** - Prohibit "Quick create" for resources. There are no comments here.
- **resource_quick_update** - Prohibit "Quick update" for resources. There are no comments here.
- **publish_document** - Prevent only resource publication, you can not cancel the publication, strange behavior, see _unpublish_document_.
- **unpublish_document** - Prevent only cancelling resource publication, you can not publish, strange behavior, see _publish_document_.
- **undelete_document** - Prevent cancellation of resource deletion. There are no comments.
- **view_document** - The section and link "Resource Review" from the context menu in the resource tree is disabled. And the list of resources is hidden from the widget. Direct link does not work. There are no comments.
- **view_unpublished** - View unpublished resources on the site. There are no comments here.
- **class_map** - Allows the user to change the Resource Type when editing / creating it. If disable, then when editing the resource, an error window pops up.
- **steal_locks** - Allows the user to intercept resource locks, in my opinion does not work.
- **edit_locked** - Allows the user to edit locked resources. In my opinion, it does not work, if the resource is open by 2 users, then for the 2nd resource is blocked.

### Environment
MODX >2.x
https://github.com/modxcms/revolution/issues/14468, https://github.com/modxcms/revolution/issues/14467, https://github.com/modxcms/revolution/issues/14436, https://github.com/modxcms/revolution/issues/14435, https://github.com/modxcms/revolution/issues/14434, https://github.com/modxcms/revolution/issues/14432, https://github.com/modxcms/revolution/issues/14431, https://github.com/modxcms/revolution/issues/14430, https://github.com/modxcms/revolution/issues/14429, https://github.com/modxcms/revolution/issues/14419, https://github.com/modxcms/revolution/issues/14405, https://github.com/modxcms/revolution/issues/14406, https://github.com/modxcms/revolution/issues/14407, https://github.com/modxcms/revolution/issues/14418

**Comment by ghost** [comment]
"Why delete and remove, is it not the same thing?"

No, "delete" means flag for deletion, "remove" means to remove the deleted item from the database.

**Comment by Ruslan-Aleev** [comment]
Are you sure? If yes, then for snippets what does it mean?
_delete_snippet - To delete or remove Snippet._

**Comment by ghost** [comment]
Yeah, for elements it is the same thing - remove from the database. Elements don't have a "delete" flag in the database like resources do.

**Comment by Ibochkarev** [comment]
Re-checked this on current 3.x (`settings_version` 3.3.0-dev). Resource keys are still split across tree CSS classes, controllers, and processors. A few 2019 bypasses are closed. `class_map` and the two lock keys are still easy to misread.

## What each key does today

| Permission | Where it is enforced | What it is *not* |
|---|---|---|
| `new_document_in_root` | `Resource/Create`, `Update` (move to root), `Duplicate`, `GetNodes` root create | Not create under a parent |
| `new_document` | Document create controller. `Resource/Create` base key. Tree / toolbar Document | Not weblink / symlink / static (those add a second key) |
| `new_symlink` / `new_weblink` / `new_static_resource` | Matching create controllers. `ActionAccessTrait` on Create. Tree `pnew_*` classes. `GetToolbar` | Opening `?a=resource/create` (that is only `new_document`) |
| `edit_document` | Update controllers (`resource/update` plus derivatives, which also need `edit_weblink` etc.) | Not Save |
| `save_document` | `Resource/Update`, `UpdateFromGrid`, `Sort`, `Reload`. Save button `canSave` | Not opening the form |
| `delete_document` | `Resource/Delete` (soft-delete). Derivatives also need `delete_weblink` / `_symlink` / `_static_resource` | Not purge (`purge_deleted`, #14467). Not Restore |
| `resource_duplicate` | `Resource/Duplicate`. Tree Duplicate. Controller `canDuplicate` | Duplicate into root still needs `new_document_in_root` |
| `resource_quick_create` / `resource_quick_update` | Tree classes `pqcreate` / `pqupdate` only | Not the full create/update pages |
| `publish_document` / `unpublish_document` | Separate processors. Tree Publish / Unpublish. `Update` checks each when the published flag changes | Lexicon for `publish_document` still says “publish or unpublish” |
| `undelete_document` | `Resource/Undelete`, `Trash/Restore`. Tree Undelete | Not purge |
| `view_document` | `resource/data`, `site_schedule`. Some list processors. Tree `pview` | Resource `Get` / `GetList` use object policy `view`, not this key |
| `view_unpublished` | Frontend `modRequest` (mgr session) | Not the manager tree |
| `class_map` | `System/Derivatives/GetList` (Resource Type combo). `Element/GetListByClass` | Not a dedicated “change type” key. Combo failure is an access-denied popup |
| `steal_locks` | `Resource/Locks/Steal`. `Resource/Update::addLock` when `steal_lock` is posted | Also needs **object** policy `steal_lock` on that resource |
| `edit_locked` | `Element/Update` only (element `locked` flag) | **Not resources.** Resource session locks are `steal_locks` |

Create processors call `checkActionPermission($classKey, 'new')`. Update calls it when `class_key` **changes**. Delete uses `'delete'`. Controllers for weblink / symlink / static create already require `new_document` **and** the type key.

## Versus the 2019 notes

**`new_document` vs type keys** — Toolbar icons are still per type (`GetToolbar`). Opening Create Resource is still `new_document` only. Changing Resource Type on that form used to bypass `new_symlink` / `new_weblink` / `new_static_resource`. **Create now checks the posted `class_key`.** Update checks a type change the same way. The 2019 bypass is closed in the processor. The combo still lists types if `class_map` is on.

**`edit_document` / `save_document`** — Still split. Form opens on edit. Save button follows `save_document`. Same as other objects after #16653.

**`delete_document`** — @wjh: for resources, delete = `deleted` flag. Remove-from-DB is `purge_deleted`. The lexicon line “delete or remove” is the same tautology as snippets. For resources the two operations really are different keys.

**`publish_document` / `unpublish_document`** — Still two keys. That is correct. The `publish_document` description is wrong.

**`view_document`** — Data page and Preview-in-tree are this key. Direct `?a=resource/data` is denied. The resource **edit** page is `edit_document`. Dashboard “list of resources” widgets vary.

**`view_unpublished`** — Front-end only, as the lexicon says.

**`class_map`** — The Resource Type combo calls `System/Derivatives/GetList`. Turn the key off: the combo errors when the panel loads. It is not a silent hide.

**`steal_locks`** — It does work if both the Administrator key **and** the Resource policy `steal_lock` are granted, and the UI posts `steal_lock`. Missing the object policy looks like “does not work.”

**`edit_locked`** — Resource “another manager has this open” is a **session lock**. That path never reads `edit_locked`. Two users on one resource is `steal_locks` / `steal_lock`. `edit_locked` is the element checkbox.

Content Editor includes `new_document`, the three type keys, `edit_document`, `save_document`, `delete_*`, `resource_duplicate`, `view_document`, `class_map`. It does **not** include `new_document_in_root`, `publish_document`, `unpublish_document`, `undelete_document`, `resource_quick_*`, `steal_locks`, `edit_locked`, `view_unpublished`.

## Suggested shape

1. Keep type keys as extras on top of `new_document` / `edit_document` / `delete_document`. Document that Create Resource + type change is now processor-enforced.
2. Fix `publish_document` lexicon (publish only).
3. Rename or document `edit_locked` vs `steal_locks` vs object `steal_lock`.
4. Either hide the Resource Type combo without `class_map` or give it a dedicated key and a quiet empty store.
5. Resource `GetList`/`Get` using object `view` while the Data page uses `view_document` is worth a short note in the lexicon.

Related: #14467 (tree / trash / `purge_deleted`), #14434 (Resource Groups tree ignores `resource_tree`).
